### ISSUE TITLE
Cannot play youtube video.

### ISSUE THREAD

**Issue by m110033** [ypchen/hdp_ims#2]
Dear YPChen,

While I was fixing the problem cannot play video on IMS, I found the link of youtube has become googlevideo as follow. I get the link of youtube by "youtube-dl".

```
# youtube-dl -g https://www.youtube.com/watch?v=oKIBJJW8neU
https://r5---sn-ipoxu-3iis.googlevideo.com/videoplayback?lmt=1471934223323770&ipbits=0&ratebypass=yes&upn=AkPvDCgBhGg&gcr=tw&source=youtube&ei=osvDV435IZ-v4ALWx6HoBQ&sver=3&key=yt6&initcwndbps=2336250&expire=1472471042&mv=m&ms=au&requiressl=yes&pl=24&itag=22&mt=1472448824&dur=314.723&id=o-AKHZ1RMcdhD_z1Fk4zH-ONpMs5rTX_Vr6ndbHwDJmVi_&mime=video%2Fmp4&sparams=dur%2Cei%2Cgcr%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Cratebypass%2Crequiressl%2Csource%2Cupn%2Cexpire&mn=sn-ipoxu-3iis&mm=31&ip=XXX.XXX.XXX.XXX&signature=B35458A6597D50FB1AE5B09A292EC87E1284B51C.65F95CB9C3F518A9B846D9C8EECA1E73FB005529
```

`$urlToGo = 'the link of youtube';`

If I send that kinds of link to "header('Location: ' . $urlToGo);" to play video on "youtube.video.php", it will shows the error message like this "無法解析的影片格式". Do you have any idea how to fixed the problem? Whetehr is the link to long to play video. Or himedia cannot parse the link without extension name?

Thank you!

**Comment by ypchen** [comment]
Hi there!

I haven't investigated the issue, but I think I have the answer to the situation you encountered.

The URL you got by executing youtube-dl is specifically for the machine on which you executed youtube-dl. In the URL, you can see that there is an "ip" parameter. I bet that's the IP of the machine, and your HiMedia machine is using another IP.

I am guessing if you put the two machines behind a NAT to share one external IP, the procedure you describe should work. As far as I can do, in most cases, the youtube URL has to be resolved on the same machine (at least the same IP), that's why youtube.video.php was written to execute on a HiMedia machine.

However, I could be wrong. It's only my educated guess based on the past experience.

Good luck.

**Comment by m110033** [comment]
Indeed, that issue of IP address I had noticed before, but I think it's not the main reason it can't play video. The device of HiMedia is behind the switch and so is my PC, so they will share the same external IP address. 

I have also tried the other way to create my own list ( A.inc, A.link.inc, ... etc ) on IMS. It will get the actual URL of youtube just like "youtube-dl" do. I have checked the value on the log of ''$fileDataURL = /tmp/yv_url.data".  Finally, it will be set to the value of $urlToGo, but it's still cannot be played.

I will keep try other ways to fix this issue. If you have any other ideas, please let me know.

Thank you!

**Comment by ypchen** [comment]
Hm... It occurs to me that there are three things you might want to check.
1. Video format. Since I see "itag=22" in the url you got. It should not be a problem. HiMedia cannot accept the new WebM formats
2. Signature. Again, since "signature=" has been in the url and looks appropriate. I believe it's not the case.
3. It's probably the SSL issue. The HiMedia built-in player software, at least for the ones I have, does not support SSL. It cannot load video data from YouTube site via SSL. If I remember correctly, after YouTube switched everything to SSL, HiMedia machines cannot play YouTube video anymore. That is why at https://github.com/ypchen/hdp_ims/commit/c85cc13771ad22f2f7b5b8ba56024aeb5ef3c4e0 the file "curl_youtube" was added to do something like

HiMedia machine <--> curl_youtube (mostly for handling SSL/HTTPS) <--> YouTube site

Later, I extended curl_youtube to also handle YouTube hls video streams. It worked for a while, but last time I checked, not working anymore.

Hence, if the URL you got can be played on your PC, I bet it's the issue that HiMedia player software (the DvdPlayer process) cannot natively handle SSL/HTTPS protocols. (I am not sure whether newer versions can do it, though.)
